### PR TITLE
Bugfix issue 665

### LINKED ISSUE #665: App crashes randomly after taking a picture with the camera
# Bug Report

## Problem
After taking a picture, pushing the OK Button (marked below) of camera-plugin crashes the app randomly.
image

### What is expected to happen?
The picture should be available in the app for further action.
 
### What does actually happen?
The app crashes.

## Information
```
2020-09-25 16:42:08.277 19807-19807/com.domain.app D/AndroidRuntime: Shutting down VM
2020-09-25 16:42:08.281 19807-19807/com.domain.app E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.domain.app, PID: 19807
    java.lang.RuntimeException: Unable to resume activity {com.domain.app/com.domain.app.MainActivity}: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=34, result=-1, data=null} to activity {com.domain.app/com.domain.app.MainActivity}: java.lang.NullPointerException: filename cannot be null
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4626)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4659)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2261)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:8107)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
     Caused by: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=34, result=-1, data=null} to activity {com.domain.app/com.domain.app.MainActivity}: java.lang.NullPointerException: filename cannot be null
        at android.app.ActivityThread.deliverResults(ActivityThread.java:5324)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4613)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4659) 
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52) 
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2261) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:237) 
        at android.app.ActivityThread.main(ActivityThread.java:8107) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100) 
     Caused by: java.lang.NullPointerException: filename cannot be null
        at android.media.ExifInterface.<init>(ExifInterface.java:1390)
        at org.apache.cordova.camera.ExifHelper.createInFile(ExifHelper.java:56)
        at org.apache.cordova.camera.CameraLauncher.processResultFromCamera(CameraLauncher.java:481)
        at org.apache.cordova.camera.CameraLauncher.onActivityResult(CameraLauncher.java:808)
        at org.apache.cordova.CordovaInterfaceImpl.onActivityResult(CordovaInterfaceImpl.java:159)
        at org.apache.cordova.CordovaActivity.onActivityResult(CordovaActivity.java:361)
        at android.app.Activity.dispatchActivityResult(Activity.java:8294)
        at android.app.ActivityThread.deliverResults(ActivityThread.java:5317)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4613) 
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4659) 
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52) 
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2261) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:237) 
        at android.app.ActivityThread.main(ActivityThread.java:8107) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(Runti
```

### Environment, Platform, Device
Chrashes have first appeared on Samsung Galaxy Tab A (SM-T515) after Android 10 update.
Same device running android 9 seams to work fine.

### Version information

Cordova Packages:

    cli: 10.0.0
        common: 4.0.2
        create: 3.0.0
        lib: 10.0.0
            common: 4.0.2
            fetch: 3.0.0
            serve: 4.0.0

Project Installed Platforms:

    android: 9.0.0

Project Installed Plugins:

    com-darryncampbell-cordova-plugin-intent: 1.3.0
    com.mirasense.scanditsdk.plugin: 5.14.4
    cordova-plugin-app-version: 0.1.9
    cordova-plugin-camera: 5.0.0 (bug appears on older versions too)
    cordova-plugin-device: 2.0.3
    cordova-plugin-file: 6.0.2
    cordova-plugin-network-information: 2.0.2
    cordova-plugin-whitelist: 1.3.4
    cordova-plugin-x-toast: 2.7.2
    phonegap-plugin-barcodescanner: 8.1.0

Environment:

    OS: Microsoft Windows 10 Pro 10.0.18362 (18362) (win32 10.0.18362) x64
    Node: v13.12.0
    npm: 6.14.4

## Checklist

- [x] I searched for existing GitHub issues
- [x] I updated all Cordova tooling to most recent version
- [x] I included all the necessary information above

### LINKED ISSUE #696: App crashes when image returned via resume event, Android
# Bug Report

## Problem
Per "android quirks" section of readme, camera may return imageData via cordova "resume" event, when Android closes main app while camera is open.  I'm triggering this scenario manually using Android settings > System > Developer Options > "Don't keep activities".  This works OK in cordova-plugin-camera 4.1.0. (imageData is returned per docs, via "resume" event.)  But in cordova-plugin-camera 5.0.1, app crashes just after taking a picture, and before/as app restores.

### What is expected to happen?
App should not crash when it restores, per scenario above.  (Works per documentation in version 4.1.0)

### What does actually happen?
App crashes after taking a picture.

## Information
- version 4.1.0 works per docs
- App doesn't appear to launch again at all.  (e.g. crash happens, and an alert("x") on the first line of JS never fires.)

### Command or Code
- Camera call I'm using:
```
	navigator.camera.getPicture(
		function (imageData) { alert("Camera callback, normal way."); },
		function (msg) { alert('Camera error: ' + msg); },  // Error CB
		{ quality:95, sourceType:Camera.PictureSourceType.CAMERA, destinationType:Camera.DestinationType.FILE_URI, correctOrientation:true, allowEdit:false, targetWidth:2000, targetHeight:2000 }
	);
```
- I'm triggering this "manually" using [Android settings > System > Developer Options > "Don't keep activities"].  That setting is important to triggering this issue.  Unless there's another way to reliably trigger the "android quirks" scenario.

### Environment, Platform, Device
Voltbuilder // Android // Sony Xperia ZX2 Compact

### Version information
Building with Voltbuilder.  I didn't specify a cordova version, so I assume they're using something recent. 

## Checklist
- [x] I searched for existing GitHub issues
- [x] I updated all Cordova tooling to most recent version
- [x] I included all the necessary information above

### PR DESCRIPTION
**Pull request by PieterVanPoyer**
### Platforms affected

Android

### Motivation and Context

Fixes #665 
Fixes #696 

### Description

In the processResultFromCamera method, the imageFilePath is used.
https://github.com/apache/cordova-plugin-camera/blob/master/src/android/CameraLauncher.java#L475 
It was not restored when the activity was destroyed.
https://github.com/apache/cordova-plugin-camera/blob/b43c78b41997294d08fc8c17a3b286a5eba5fa6d/src/android/CameraLauncher.java#L1356
This resulted in a NullPointerException because the resulting sourcePath was null.
https://github.com/apache/cordova-plugin-camera/blob/b43c78b41997294d08fc8c17a3b286a5eba5fa6d/src/android/CameraLauncher.java#L473

So I've added the imageFilePath to the state Bundle in the saveInstance and the restoreInstance methods.

You can install and test the plugin (on Android) with next command.

````
npx cordova plugin remove cordova-plugin-camera
npx cordova plugin add https://github.com/PieterVanPoyer/cordova-plugin-camera/#bugfix/issue-665-save-instance-restore-bug
````

You could test it with next reproduction repo:
https://github.com/PieterVanPoyer/cordova-camera-plugin-testing-app

### Testing

Tested it on an Android 10 emulator.
I enabled the developers mode on the emulator and made the activity destroy everytime the camera is shown.
So I forced the resume flow on Android

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [x] I added automated test coverage as appropriate for this change
- [ ] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [x] I've updated the documentation if necessary

### CODE REVIEW

**Comment by PieterVanPoyer** [comment]
@breautek @jcesarmobile @erisu Is anybody able to review this PR? Kind regards.

**Review by breautek — APPROVED** [review]
lgtm

**Comment by PieterVanPoyer** [comment]
@breautek @erisu maybe it's time to merge this one. And start a release. Some users report that this PR fixes the linked issues.

File: src/android/CameraLauncher.java [hunk 1]
```diff
@@ -1350,6 +1351,10 @@ public Bundle onSaveInstanceState() {
             state.putString(IMAGE_URI_KEY, this.imageFilePath);
         }
 
+        if(this.imageFilePath != null) {
```
Comment (timbru31) [inline src/android/CameraLauncher.java:1354]:
```suggestion
        if (this.imageFilePath != null) {
```

**Comment by PieterVanPoyer** [comment]
@timbru31 thanks for the review. I did commit your suggestion.

**Comment by PieterVanPoyer** [comment]
@timbru31 or @breautek can anyone merge this one to the master. Most users report that this PR fixes the linked issues.

**Comment by breautek** [comment]
Generally I like to have 2 green checks, but I'll merge sometime Monday assuming there are no objections. Sooner if I get another green check from a PMC member.

**Comment by breautek** [comment]
Tests are green, and haven't had any objections so merging!

Thank you for your contribution @PieterVanPoyer

**Comment by FlossyWeb** [comment]
Hi there,
Are you planning to update the npm package anytime soon ?
It would be nice...
Thank you guys

**Comment by codeconsole** [comment]
Any chance for a merge or at least a comment on https://github.com/apache/cordova-plugin-camera/pull/712 ?

**Comment by satheeshkumartitan** [comment]
I've also updated the latest plugin 5.0.2, but the same problems occur on Android 10 (device - redme not 7pro). The plugin 5.0.2. Forcefully close the application once we capture a photo. This happens on some specific devices. Please update the new version I'm waiting for the new version 
Node -v ----14.16.0
ionic -v  ---- 6.13.1
cordova -v----10.0.0
ionic cordova platform add android@9.1.0
tested  <preference name="android-targetSdkVersion" value="30" /> and  <preference name="android-targetSdkVersion" value="29" /> both same crash.


I'm waiting for your response, thanks

**Comment by jay34fr** [comment]
Hi all,
Same issue here.
I updated to 5.0.2 but the app is still restarting after taking and validating a photo.

Cordova 9.0.0
Cordova Android 9.0.0

**Pull request merged**